File: LR1/Array_Lists_Stack/SinglyLinkedList.cpp
#include "SinglyLinkedList.h"

// Добавление в начало
void FaddFront(FList*& nd, string value) {
    FList* newNode = new FList;    // создаем новый узел
    newNode->key = value;
    newNode->next = nd;        // новый указывает на старый "первый"
    nd = newNode;              // теперь он — первый
}

// Добавление в конец
void FaddBack(FList*& nd, string value) {
    FList* newNode = new FList;
    newNode->key = value;
    newNode->next = nullptr;

    if (!nd) {
        nd = newNode;
        return;
    }

    FList* temp = nd;
    while (temp->next)
        temp = temp->next;
    temp->next = newNode;
}

// --- Добавление после узла ---
bool FaddAfter(FList* nd, string value) {
    if (!nd) return false;

    FList* newNode = new FList{ value, nd->next };
    nd->next = newNode;
    return true;
}

// --- Добавление перед узлом ---
bool FaddBefore(FList*& head, FList* nd, string value) {
    if (!head || !nd) return false;

    // если добавляем перед головой
    if (head == nd) {
        FaddFront(head, value);
        return true;
    }

    FList* temp = head;
    while (temp->next && temp->next != nd)
        temp = temp->next;

    if (!temp->next) return false; // nd не найден

    FList* newNode = new FList{ value, nd };
    temp->next = newNode;
    return true;
}

// Удаление первого элемента
bool FreFront(FList*& nd) {
    if (!nd) return false; // список пуст

    FList* tempDel = nd;
    nd = nd->next;
    delete tempDel;
    return true;
}

// Удаление последнего элемента
bool FreBack(FList*& nd) {
    if (!nd) return false; // пустой список

    if (!nd->next) {
        delete nd;
        nd = nullptr;
        return true;
    }

    FList* temp = nd;
    while (temp->next->next)
        temp = temp->next;

    delete temp->next;
    temp->next = nullptr;
    return true;
}

// --- Удаление после узла ---
bool FreAfter(FList* nd) {
    if (!nd || !nd->next) return false;

    FList* temp = nd->next;
    nd->next = temp->next;
    delete temp;
    return true;
}

// --- Удаление перед узлом ---
bool FreBefore(FList*& head, FList* nd) {
    if (!head || !nd || head == nd) return false;

    // если нужно удалить сам head->next перед nd
    if (head->next == nd) {
        FList* temp = head;
        head = nd;
        delete temp;
        return true;
    }

    FList* temp = head;
    while (temp->next && temp->next->next != nd)
        temp = temp->next;

    if (!temp->next) return false; // nd не найдено

    FList* toDelete = temp->next;
    temp->next = nd;
    delete toDelete;
    return true;
}

// Удаление по значению
bool FreValue(FList*& nd, string value) {
    if (!nd) return false;

    if (nd->key == value) {
        FList* tempDel = nd;
        nd = nd->next;
        delete tempDel;
        return true;
    }

    FList* cur = nd;
    while (cur->next && cur->next->key != value)
        cur = cur->next;

    if (!cur->next) return false;

    FList* tempDel = cur->next;
    cur->next = cur->next->next;
    delete tempDel;
    return true;
}

// Поиск по значению
FList* Fsearch(FList* nd, string value) {
    FList* cur = nd;
    while (cur) {
        if (cur->key == value)
            return cur;
        cur = cur->next;
    }
    return nullptr;
}

// Вывод списка
void print(FList* nd) {
    FList* cur = nd;
    while (cur) {
        cout << cur->key << " ";
        cur = cur->next;
    }
    //cout << "nullptr";
}

// --- Рекурсивный вывод ---
void printRecursive(FList* head) {
    if (!head) {
        //cout << "nullptr\n";
        return;
    }
    cout << head->key << " ";
    printRecursive(head->next);
}

// Очистка памяти
void Fclear(FList*& nd) {
    while (nd) {
        FList* temp = nd;
        nd = nd->next;
        delete temp;
    }
}

File: LR1/NaGO/Array_Lists_Stack/structures/singlylist/singly_linked_list.go
package singlylist

import "fmt"

// FList — элемент односвязного списка
type FList struct {
	Key  string
	Next *FList
}

// Добавление в начало
func FaddFront(head **FList, value string) {
	newNode := &FList{Key: value, Next: *head}
	*head = newNode
}

// Добавление в конец
func FaddBack(head **FList, value string) {
	newNode := &FList{Key: value, Next: nil}
	if *head == nil {
		*head = newNode
		return
	}
	temp := *head
	for temp.Next != nil {
		temp = temp.Next
	}
	temp.Next = newNode
}

// Добавление после узла
func FaddAfter(nd *FList, value string) bool {
	if nd == nil {
		return false
	}
	newNode := &FList{Key: value, Next: nd.Next}
	nd.Next = newNode
	return true
}

// Добавление перед узлом
func FaddBefore(head **FList, nd *FList, value string) bool {
	if *head == nil || nd == nil {
		return false
	}
	if *head == nd {
		FaddFront(head, value)
		return true
	}
	temp := *head
	for temp.Next != nil && temp.Next != nd {
		temp = temp.Next
	}
	if temp.Next == nil {
		return false
	}
	newNode := &FList{Key: value, Next: nd}
	temp.Next = newNode
	return true
}

// Удаление первого элемента
func FreFront(head **FList) bool {
	if *head == nil {
		return false
	}
	*head = (*head).Next
	return true
}

// Удаление последнего элемента
func FreBack(head **FList) bool {
	if *head == nil {
		return false
	}
	if (*head).Next == nil {
		*head = nil
		return true
	}
	temp := *head
	for temp.Next.Next != nil {
		temp = temp.Next
	}
	temp.Next = nil
	return true
}

// Удаление после узла
func FreAfter(nd *FList) bool {
	if nd == nil || nd.Next == nil {
		return false
	}
	nd.Next = nd.Next.Next
	return true
}

// Удаление перед узлом
func FreBefore(head **FList, nd *FList) bool {
	if *head == nil || nd == nil || *head == nd {
		return false
	}
	if (*head).Next == nd {
		*head = nd
		return true
	}
	temp := *head
	for temp.Next != nil && temp.Next.Next != nd {
		temp = temp.Next
	}
	if temp.Next == nil {
		return false
	}
	temp.Next = nd
	return true
}

// Удаление по значению
func FreValue(head **FList, value string) bool {
	if *head == nil {
		return false
	}
	if (*head).Key == value {
		*head = (*head).Next
		return true
	}
	cur := *head
	for cur.Next != nil && cur.Next.Key != value {
		cur = cur.Next
	}
	if cur.Next == nil {
		return false
	}
	cur.Next = cur.Next.Next
	return true
}

// Поиск по значению
func Fsearch(nd *FList, value string) *FList {
	cur := nd
	for cur != nil {
		if cur.Key == value {
			return cur
		}
		cur = cur.Next
	}
	return nil
}

// Вывод списка
func Print(nd *FList) {
	cur := nd
	for cur != nil {
		fmt.Print(cur.Key, " ")
		cur = cur.Next
	}
	fmt.Println()
}

// Рекурсивный вывод
func PrintRecursive(head *FList) {
	if head == nil {
		return
	}
	fmt.Print(head.Key, " ")
	PrintRecursive(head.Next)
}

// Очистка списка
func Fclear(head **FList) {
	for *head != nil {
		*head = (*head).Next
	}
}
